src/libraries/System.Private.Uri/src/System/UriExt.cs

```diff
@@ -213,22 +212,24 @@ private void InitializeUri(ParsingError err, UriKind uriKind, out UriFormatExcep
             }
         }
 
-        //
         // Unescapes entire string and checks if it has unicode chars
-        //
-        private bool CheckForUnicode(string data)
+        // Also checks for sequences that are 3986 Unreserved characters as these should be un-escaped
+        private static bool CheckForUnicodeOrEscapedUnreserved(string data)
         {
             for (int i = 0; i < data.Length; i++)
             {
                 char c = data[i];
                 if (c == '%')
                 {
-                    if (i + 2 < data.Length)
+                    if ((uint)(i + 2) < (uint)data.Length)
```

Comment:
Same question as on the other PR; does this uint casting actually enable the JIT to eliminate the i+1 and i+2 bounds checks?  If it does, great; I just didn't think it did.

Reply:
I measured it against a completely unsafe version and the perf is: (`new` being the unsafe variant) 

|         Method |              Toolchain |     Mean |   Error |  StdDev |   Median | Ratio | RatioSD |
|--------------- |----------------------- |---------:|--------:|--------:|---------:|------:|--------:|
|          Ascii | \perf-base\CoreRun.exe | 157.4 ns | 0.26 ns | 1.33 ns | 157.0 ns |  0.95 |    0.01 |
|          Ascii |  \perf-new\CoreRun.exe | 165.7 ns | 0.34 ns | 1.68 ns | 165.5 ns |  1.00 |    0.00 |
|                |                        |          |         |         |          |       |         |
| PercentEncoded | \perf-base\CoreRun.exe | 192.7 ns | 1.94 ns | 9.67 ns | 186.9 ns |  0.98 |    0.05 |
| PercentEncoded |  \perf-new\CoreRun.exe | 196.7 ns | 0.25 ns | 1.22 ns | 196.6 ns |  1.00 |    0.00 |

That said, I'll merge it as-is